Rigged hand visualizer

rolandsmeenk opened the pull request:
## Overview

Implemented a RiggedHandVisualizer that can be used to visualize a skinned mesh for hands.

![image](https://user-images.githubusercontent.com/480316/83201369-a3bb5500-a145-11ea-867d-31d56196d5df.png)

Implementation is based on and has visual parity with the LeapMotion RiggedHand/RiggedFinger combi. Fixed an issue in the LeapMotionArticulatedHand where the wrong thumb metacarpal index was copied from the leapmotion data. The handmodels are a customization of the LeapMotion handmodels with a bit rounder shapes and a different material.
Cleaner handmodels may be needed, but my rigging skills are lacking.
Tested in editor with Leapmotion and on HoloLens 2.

## Gifs
| WMR Headset + Leap | Input Simulation |  
|--------|-------|
|    ![RiggedHandsLeapVR](https://user-images.githubusercontent.com/53493796/86183350-a6172300-bae6-11ea-917e-b2a001dc110a.gif)| ![RiggedHandsInputSimulation](https://user-images.githubusercontent.com/53493796/86183368-aca59a80-bae6-11ea-8e60-c35567b31e38.gif)|  

## Changes

Related to #4257

rolandsmeenk commented:
@CDiaz-MS I cleaned up a little too much which caused the meshes to be empty. I reimported the models that are used by the prefabs and restored the connections. I should work now.

rolandsmeenk commented:
@CDiaz-MS Ok, replaced with FBX versions. Hopefully you will some hands now. :)

CDiaz-MS commented:
The .blend -> .fbx worked.  I am seeing the hands!

CDiaz-MS commented:
@rolandsmeenk  Added some gifs to the PR description so people can see it in action

CDiaz-MS commented on Assets/MRTK/SDK/Experimental/RiggedHandVisualizer/RiggedHandVisualizer.cs:
```diff
@@ -0,0 +1,254 @@
+    public class RiggedHandVisualizer : MonoBehaviour, IMixedRealityControllerVisualizer, IMixedRealitySourceStateHandler, IMixedRealityHandJointHandler
+    {
+        public virtual Handedness Handedness { get; set; }
+
+        public GameObject GameObjectProxy => gameObject;
+
+        public IMixedRealityController Controller { get; set; }
+
+        public Transform Wrist;
+        public Transform Palm;
```
I have highlighted a couple here but doc comments are useful for public properties, [mrtk coding guidelines](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/CodingGuidelines.html#function--method-summary-headers)

CDiaz-MS commented on Assets/MRTK/SDK/Experimental/RiggedHandVisualizer/RiggedHandVisualizer.cs:
```diff
@@ -0,0 +1,254 @@
+        public Transform RingMetacarpal;
+        public Transform PinkyMetacarpal;
+
+        [Tooltip("Hands are typically rigged in 3D packages with the palm transform near the wrist. Uncheck this if your model's palm transform is at the center of the palm similar to Leap API hands.")]
+        public bool modelPalmAtLeapWrist = true;
+
+        [Tooltip("Allows the mesh to be stretched to align with finger joint positions. Only set to true when mesh is not visible.")]
+        public bool deformPosition = true;
+
```
Public properties should be capitalized, also applies to the properties below, [naming guidelines mrtk](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/CodingGuidelines.html#naming-conventions)

CDiaz-MS commented on Assets/MRTK/SDK/Experimental/RiggedHandVisualizer/RiggedHandVisualizer.cs:
```diff
@@ -0,0 +1,254 @@
+                joints[TrackedHandJoint.PinkyDistalJoint] = PinkyMetacarpal.GetChild(0).GetChild(0).GetChild(0);
+                joints[TrackedHandJoint.PinkyTip] = PinkyMetacarpal.GetChild(0).GetChild(0).GetChild(0).GetChild(0);
+            }
+
+            // Precalculated values for LeapMotion testhand
+            FingerTipLengths[TrackedHandJoint.ThumbTip] = 0.02167f;
+            FingerTipLengths[TrackedHandJoint.IndexTip] = 0.01582f;
+            FingerTipLengths[TrackedHandJoint.MiddleTip] = 0.0174f;
+            FingerTipLengths[TrackedHandJoint.RingTip] = 0.0173f;
+            FingerTipLengths[TrackedHandJoint.PinkyTip] = 0.01596f;
```
It's best to avoid seemingly magic numbers, it's helpful to store the pre-calculated values in another var with a descriptive name.

provencher commented:
Would be nice to have a way to easily tie this into the HandMesh bool found in the HandTracking profile. 

Also, I'd love to see the distance between index and thumb used to create a visual feedback in the material to signal pinch strength.

Otherwise nice work! This is a great contribution!

rolandsmeenk commented:
> Would be nice to have a way to easily tie this into the HandMesh bool found in the HandTracking profile.

Do you mean the HandMesh Prefab? I looked at adding this into the BaseHandVisualizer, but chose to not add more functionality to that and implement a separate visualizer for now. Maybe there can be a composite hand visualizer where a user can mix joint prefabs, hand mesh and a skinnedmesh to their liking.

> Also, I'd love to see the distance between index and thumb used to create a visual feedback in the material to signal pinch strength.

Pinch strength visualization would be indeed be a nice feature, but I would prefer to keep that outside of this PR. I can imagine this would also be useful on the dynamic handmesh.

CDiaz-MS commented:
/azp run

CDiaz-MS commented:
Looks like this file is missing a namespace declaration:
```
WARNING: f:\aipmragent_work\3\s\Assets\MRTK\SDK\Experimental\RiggedHandVisualizer\RiggedHandVisualizer.cs is missing a 
namespace declaration (i.e. missing namespace Microsoft.MixedReality.Toolkit.*
```

CDiaz-MS commented:
/azp run

rolandsmeenk commented:
Ok, I'll cook something up

CDiaz-MS commented:
/azp run

CDiaz-MS commented:
@rolandsmeenk Looks like docs validation did not pass, there is an invalid file link in the documentation
```
[20-06-09 10:40:15.769]Warning:[BuildCommand.BuildCore.Build Document.LinkPhaseHandlerWithIncremental.ConceptualDocumentProcessor.Save](Assets/MRTK/SDK/Experimental/RiggedHandVisualizer/README_RiggedHandVisualizer.md#L15)Invalid file link:(~/Assets/MRTK/Examples/Experimental/RiggedHandVisualizer/Scenes).

```

rolandsmeenk commented:
I do not understand exactly why this goes wrong. I moved the example scene into a separate folder to be in lign with the other examples. The documentation is linking to that scene folder with a relative path and that works in my branch. It's not really a file link, but a link to the folder that contains the example.

CDiaz-MS commented:
/azp run

davidkline-ms commented on Assets/MRTK/SDK/Experimental/RiggedHandVisualizer/Prefabs/hand_left.fbx.meta:
```diff
@@ -0,0 +1,163 @@
+fileFormatVersion: 2
+guid: 5d3838939787e724fa3eaf6a2024f126
```
@rolandsmeenk , what is the source of the hand models? did you create them?

CDiaz-MS commented:
@rolandsmeenk Unfortunately we cannot merge the hand models used in this PR because they belong to Ultraleap.  Can you try using the existing hand models in MRTK from this PR:
https://github.com/microsoft/MixedRealityToolkit-Unity/pull/6943/files

rolandsmeenk commented:
@CDiaz-MS I tried the handcoach hands before, but they are rigged a bit differently and the results do not look good. I was thinking of extracting a handmesh from HoloLens 2 and trying to rig, but that is still WIP.

rolandsmeenk commented:
@CDiaz-MS I replaced the hand models with hands that are a modification of the handcoach models. Sliced off a piece of the arm, adjust some weight for the wrist and fixed a rotation in the thumb. Deformation of joints and the metacarpals are not used in this setup so the hands do not adjust to the size of the tracked hand, because that looks better with this rig. I am still not entirely happy with how the fingers are still curved when you stretch out your hand.

isrg-marvel commented:
@rolandsmeenk - This is great work. Instead of UltraLeap/LeapMotion Sensors, Would the RiggedHandVisualizer.cs work with the the HandTracking data from Hololens-2?

CDiaz-MS commented:
@rolandsmeenk Tested this out and I cannot see the new hands.  I saw that you removed the modified leap motion rigged hands, and you mentioned that there should be rigged hands based on the MRTK hand coach models.  Are the new rigged hands missing?

rolandsmeenk commented:
@CDiaz-MS Forgot to check in the adjusted hand models.

rolandsmeenk commented:
@isrg-marvel Yes, this works with the HoloLens 2 too. This is an alternative for the BaseHandVisualizer based hand visualization.

CDiaz-MS commented:
@rolandsmeenk I cannot see the left hand when I test this with leap in editor.  The RiggedHandRight prefab does not have a runtime animation controller while the left does 

![RiggedHandsPR2](https://user-images.githubusercontent.com/53493796/85908833-a607e200-b7cb-11ea-98bb-97aec335500a.gif)

rolandsmeenk commented:
@CDiaz-MS Mmm, from what you describe I would either expect that the "Global Left Hand Visualizer" in the  controller mapping was not set or that the prefab would not have "Update when offscreen" enabled on the SkinnedMeshRenderer. Both are not the case. I made a fresh clone, but both hands show up with LeapMotion in the editor. Did a rebase of the branch, but I did not see any issues there. Can you see the RiggedHandLeft and RiggedHandRight instances appear in the MixedRealityPlaySpace when a hand is detected by LeapMotion? In my experience I sometimes need to reenter the LeapMotion FOV to get both of my hands recognized (LM in desktop setup). Otherwise I don't know what could cause this.

CDiaz-MS commented:
/azp run

CDiaz-MS commented:
@rolandsmeenk Tested again and both of the hands are showing up for me.  I also added some updated gifs to the description with the new hand models.
